### PR TITLE
HPCC-20145 Make soapplus stress test work with persistent connections

### PR DESCRIPTION
**Pull request by mayx**
- When persistent connection is not enabled, send "Connection: close"
  header to server so that the connection will be closed
- When persistent connection is enabled, rely on Content-Length instead of
  EOF to read the response content.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by mayx** [comment]
@mckellyln Mark please review...

**Comment by mckellyln** [comment]
This test -

soapplus -url https://.:19876/roxie -stress 20 4 -persist -delay 4 6 -d 1 -i echo6.xml -vx

produces -

000000C1 2018-09-04 10:11:29.230  2666 21727 "  /opt/HPCCSystems/lib/libjlib.so(_Z16printStackReportx+0x4f) [0x7ff00a603f1f]"
000000C2 2018-09-04 10:11:29.230  2666 21727 "  /opt/HPCCSystems/lib/libjlib.so(_Z20raiseAssertExceptionPKcS0_j+0x93) [0x7ff00a604283]"
000000C3 2018-09-04 10:11:29.230  2666 21727 "  /opt/HPCCSystems/lib/libthorhelper.so(_ZN11CSafeSocket9readBlockER12StringBufferjP10HttpHelperRbS4_j+0x588) [0x7ff00ca8af98]"
000000C4 2018-09-04 10:11:29.230  2666 21727 "  /opt/HPCCSystems/lib/libccd.so(_ZN17RoxieSocketWorker6doMainEPKc+0x2c1) [0x7ff00d955521]"
000000C5 2018-09-04 10:11:29.230  2666 21727 "  /opt/HPCCSystems/lib/libjlib.so(_ZN20CPooledThreadWrapper3runEv+0x9f) [0x7ff00a6c6c9f]"
000000C6 2018-09-04 10:11:29.230  2666 21727 "  /opt/HPCCSystems/lib/libjlib.so(_ZN6Thread5beginEv+0x28) [0x7ff00a6bee78]"
000000C7 2018-09-04 10:11:29.230  2666 21727 "  /opt/HPCCSystems/lib/libjlib.so(_ZN6Thread11_threadmainEPv+0x1f) [0x7ff00a6c0a9f]"
000000C8 2018-09-04 10:11:29.230  2666 21727 "  /lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba) [0x7ff009f036ba]"
000000C9 2018-09-04 10:11:29.230  2666 21727 "  /lib/x86_64-linux-gnu/libc.so.6(clone+0x6d) [0x7ff00d00a41d]"
000000CA 2018-09-04 10:11:29.230  2666 21727 "ERROR: assert(bytesRead == sizeof(len)) failed - file: roxiehelper.cpp, line 1479"

See also https://track.hpccsystems.com/browse/HPCC-20369

**Comment by mayx** [comment]
@mckellyln The core you posted seems to be on the server side, we'll need to look at it of course, but better in a different ticket maybe?

**Comment by mayx** [comment]
@mckellyln I'm not able to reproduce the core, with the same test input and parameters as you provided...

**Comment by mckellyln** [comment]
With HPCC-20369 merged and this PR I now see success:
```
+ /opt/HPCCSystems/bin/soapplus -url http://.:9876/roxie -d 1 -i echo6.xml -vx
Successfully parsed XML

---------------
Number of Threads:                1
Total hits:                       1
Run length(millisecond):          1
Hits per second:                  1000.0
Total data sent:                  1677
Total data received:              2123
Total data transferred:           3800
Data transferred per second:      3800000.0
Slowest round trip(millisecond):  1
Fastest round trip(millisecond):  1
Average round trip(millisecond):  1

---------------
+ /opt/HPCCSystems/bin/soapplus -url https://.:19876/roxie -d 1 -i echo6.xml -vx
Successfully parsed XML

---------------
Number of Threads:                1
Total hits:                       1
Run length(millisecond):          2
Hits per second:                  500.0
Total data sent:                  1678
Total data received:              2123
Total data transferred:           3801
Data transferred per second:      1900500.0
Slowest round trip(millisecond):  2
Fastest round trip(millisecond):  2
Average round trip(millisecond):  2

---------------
+ /opt/HPCCSystems/bin/soapplus -url http://.:9876/roxie -stress 20 4 -delay 4 6 -d 1 -i echo6.xml -vx
Started 20 stress test threads.
Running the test for 4 seconds...
All 20 stresstest threads finished.

---------------
Number of Threads:                20
Total hits:                       13081
Run length(millisecond):          4000
Hits per second:                  3270.2
Total data sent:                  21936837
Total data received:              27770963
Total data transferred:           49707800
Data transferred per second:      12426950.0
Slowest round trip(millisecond):  19
Fastest round trip(millisecond):  0
Average round trip(millisecond):  1

---------------
+ /opt/HPCCSystems/bin/soapplus -url http://.:9876/roxie -stress 20 4 -delay 4 6 -persist -d 1 -i echo6.xml -vx
Started 20 stress test threads.
Running the test for 4 seconds...
All 20 stresstest threads finished.

---------------
Number of Threads:                20
Total hits:                       12042
Run length(millisecond):          4000
Hits per second:                  3010.5
Total data sent:                  19965636
Total data received:              25825422
Total data transferred:           45791058
Data transferred per second:      11447764.5
Slowest round trip(millisecond):  24
Fastest round trip(millisecond):  0
Average round trip(millisecond):  1

---------------
+ /opt/HPCCSystems/bin/soapplus -url https://.:19876/roxie -stress 20 4 -delay 4 6 -d 1 -i echo6.xml -vx
Started 20 stress test threads.
Running the test for 4 seconds...
All 20 stresstest threads finished.

---------------
Number of Threads:                20
Total hits:                       2300
Run length(millisecond):          4000
Hits per second:                  575.0
Total data sent:                  3859400
Total data received:              4882900
Total data transferred:           8742300
Data transferred per second:      2185575.0
Slowest round trip(millisecond):  37
Fastest round trip(millisecond):  5
Average round trip(millisecond):  30

---------------
+ /opt/HPCCSystems/bin/soapplus -url https://.:19876/roxie -stress 20 4 -delay 4 6 -persist -d 1 -i echo6.xml -vx
Started 20 stress test threads.
Running the test for 4 seconds...
All 20 stresstest threads finished.

---------------
Number of Threads:                20
Total hits:                       11832
Run length(millisecond):          4000
Hits per second:                  2958.0
Total data sent:                  19629288
Total data received:              25375080
Total data transferred:           45004368
Data transferred per second:      11251092.0
Slowest round trip(millisecond):  63
Fastest round trip(millisecond):  0
Average round trip(millisecond):  1
```

File: esp/tools/soapplus/http.cpp [hunk 1]
```diff
@@ -1229,40 +1268,130 @@ int HttpClient::sendStressRequest(StringBuffer& request, HttpStat* stat)
...
-        int len = 1;
+        int len = 0;
         char recvbuf[2048];
-        StringBuffer xml;
-        while(1)
+        if (!isPersistent)
+        {
+            while (1)
+            {
+                len = sock->receive(recvbuf, 2047);
```
Comment (mckellyln) [inline esp/tools/soapplus/http.cpp]: This wont return now until either 2047 bytes are sent or the connection is closed.  Is that always ok ?

**Comment by mayx** [comment]
@mckellyln great! Are you ready to approve this PR? :)

File: esp/tools/soapplus/http.cpp [hunk 2]
```diff
@@ -1229,40 +1268,130 @@ int HttpClient::sendStressRequest(StringBuffer& request, HttpStat* stat)
...
+            unsigned int searchStart = 0;
+            while (1)
             {
-                total_len += len;
-                recvbuf[len] = 0;
-                if (m_doValidation)
-                    xml.append(len, recvbuf);
-                if(http_tracelevel >= 10)
-                    fprintf(m_logfile, "%s", recvbuf);
+                len = sock->receive(recvbuf, 0, 2047);
```
Comment (mckellyln) [inline esp/tools/soapplus/http.cpp]: Should min be 0 here ?  Maybe does not matter but not sure.

**Comment by mckellyln** [comment]
@mayx one question about the length.  Could it return 0 when it should have kept trying for the header ?
I am not informed about the http content parsing, but overall the rest looks good.

File: esp/tools/soapplus/http.cpp (hunk 2 above)
Comment (mayx) [reply esp/tools/soapplus/http.cpp]: This shouldn't matter. Socket will block until either data arrives or connection is closed or error happens, in another word, there's no case where it just receives 0 byte without connection close or error. Hard to tell the minimum header length, but 0 is safe here.

**Comment by mayx** [comment]
@mckellyln no it won't return 0 when it should have kept trying for the header...

File: esp/tools/soapplus/http.cpp (hunk 2 above)
Comment (mayx) [reply esp/tools/soapplus/http.cpp]: @mckellyln I looked at the code further. 0 is fine unless EINTR happens, so the min is better to be 1 here.

Comment (mayx) [reply esp/tools/soapplus/http.cpp]: Actually the ENITR issue can be avoided by adding a "continue" there. But I'll still change 0 to 1 as I do expect some data (at least 1). I'll also add the continue, as the rest of the loop is unnecessary for that case.

**Comment by mayx** [comment]
@mckellyln change pushed. My tests worked. Please review and run your tests again...

**Comment by mckellyln** [comment]
@mayx I am getting this error now, with latest master -
```
SSL_accept returned -1, SSL_get_error=1, error - error:140760FC:SSL routines:SSL23_GET_CLIENT_HELLO:unknown protocol"
```
So we need to debug that before getting back to your latest change.

**Comment by mckellyln** [comment]
@mayx test ran ok.  Looks good.

**Review by mckellyln — APPROVED** [review]
good to merge

**Pull request merged**